fhir-server/src/main/java/au/csiro/pathling/fhirpath/FhirPath.java

```diff
@@ -62,34 +62,36 @@
   boolean isSingular();
 
   /**
-   * Returns the specified child of this path, if there is one.
+   * Creates a copy of this FhirPath with an updated {@link Dataset}, ID and value {@link Column}s.
    *
-   * @param name The name of the child element
-   * @return An {@link ElementDefinition} object
+   * @param expression an updated expression to describe the new FhirPath
+   * @param dataset the new Dataset that can be used to evaluate this FhirPath against data
+   * @param idColumn the new resource identity column
+   * @param valueColumn the new expression value column
+   * @param singular the new singular value
+   * @param thisColumn a column containing the collection being iterated, for cases where a path is
+   * being created to represent the {@code $this} keyword
+   * @return a new instance of FhirPath
    */
   @Nonnull
-  Optional<ElementDefinition> getChildElement(@Nonnull String name);
+  FhirPath copy(@Nonnull String expression, @Nonnull Dataset<Row> dataset,
+      @Nonnull Optional<Column> idColumn, @Nonnull Column valueColumn, boolean singular,
+      @Nonnull Optional<Column> thisColumn);
 
   /**
-   * Returns the resource value column from the resource at the root of this path.
-   * <p>
-   * This is required for reverse reference resolution, where we need to get to the resource to be
-   * joined to the source of the reverse resolve operation.
+   * Gets an ID {@link Column} from any of the inputs, if there is one.
    *
-   * @return A {@link Column}
+   * @param inputs a collection of objects
+   * @return a {@link Column}, if one was found
    */
   @Nonnull
-  Optional<Column> getOriginColumn();
-
-  /**
-   * Returns the resource type of the resource at the root of this path.
-   * <p>
-   * This is required for reverse reference resolution, where we need to get to the resource to be
-   * joined to the source of the reverse resolve operation.
-   *
-   * @return A {@link ResourceDefinition}
-   */
-  @Nonnull
-  Optional<ResourceDefinition> getOriginType();
+  static Optional<Column> findIdColumn(@Nonnull final Object... inputs) {
+    return Stream.of(inputs)
```

Comment:
More as a curiosity - the first two lines can be also written as:
```
    return Stream.of(inputs)
        .filter(FhirPath.class::isInstance)
        .map(FhirPath.class::cast)
```

Reply:
Now that I am reading about the difference between `instanceof` and `isInstance`, I am wondering how `instanceof` works at all for what we are using it for.

It sounds like `instanceof` is a compile-time only thing...

Reply:
... which can't be right, because it _does_ work to interrogate types at runtime. We have tests for this.

Reply:
Ok, learned scholar @dionmcm informs me that the two main differences are:

- `instanceof` provides a level of compile-time checking
- `isInstance` needs to be used when the class handle is not static

In light of this, I propose we use `instanceof` when we are dealing with static class identifiers, so that we can get the benefit of the compile-time checks.